sdk/component/env.go

```diff
@@ -0,0 +1,123 @@
+package component
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"strconv"
+	"time"
+)
+
+type (
+	// parseableEnvTypes represents the types the parser is capable of handling.
+	// TODO: extend with slices if needed.
+	parseableEnvTypes interface {
+		string | bool | int | uint | int64 | uint64 | float64 | time.Duration | time.Time | url.URL
+	}
+
+	// envLoader is an alias for a function that loads values from the env. It mirrors the signature of os.Getenv.
+	envLoader func(key string) string
+
+	envParseOpts struct {
+		envLoader      envLoader
+		defaultOnError bool
+		timeLayout     string
+		sensitive      bool
+	}
+
+	// envParseOption is a means to customize parse options via variadic parameters.
+	envParseOption func(o *envParseOpts) error
+)
+
+var (
+	defaultEnvParseOptions = envParseOpts{
+		envLoader:      os.Getenv,
+		defaultOnError: false,
+		timeLayout:     time.RFC3339,
+	}
+)
+
+// withEnvLoader allows overriding how env vars are loaded.
+//
+// Primarily used for testing.
+func withEnvLoader(loader envLoader) envParseOption {
+	return func(o *envParseOpts) error {
+		if loader == nil {
+			return errors.New("env loader function cannot be nil")
+		}
+
+		o.envLoader = loader
+		return nil
+	}
+}
+
+// withFallbackToDefaultOnError informs the parser that if an error is encountered during parsing, it should fallback to the default value.
+func withFallbackToDefaultOnError(fallback bool) envParseOption {
+	return func(o *envParseOpts) error {
+		o.defaultOnError = fallback
+		return nil
+	}
+}
+
+// fromEnvOrDefault attempts to parse the environment variable provided. If it is empty or missing, the default value is used.
+//
+// If an error is encountered, depending on whether the `withFallbackToDefaultOnError` option is provided it will either
+// fallback or return the error back to the client.
+func fromEnvOrDefault[T parseableEnvTypes](envVar string, defaultVal T, opts ...envParseOption) (dest T, err error) {
```

Comment:
This handles any type of env var in the same way and we can tell it to error if an env var is required, otherwise a default can be used.

This approach has been battle tested at Cloudflare.